compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/TextFieldDelegate.skiko.kt

```diff
@@ -0,0 +1,158 @@
+        caretOffsetPosition = textLayoutResult.value.getLineStart(lineNumber)
+    } else if (textLayoutResult.isRightEdgeTapped(offset)) {
+        val lineNumber = textLayoutResult.value.getLineForOffset(offset)
+        caretOffsetPosition = textLayoutResult.value.getLineEnd(lineNumber)
+    } else if (textLayoutResult.isPunctuationOrSpaceBeforeWordTapped(offset, currentText)) {
+        val nextWordBoundary = findNextWordBoundary(offset, currentText, textLayoutResult)
+        caretOffsetPosition = nextWordBoundary.start
+    } else if (textLayoutResult.isFirstLetterOfWordTapped(offset)) {
+        val wordBoundary = textLayoutResult.value.getWordBoundary(offset)
+        caretOffsetPosition = wordBoundary.start
+    } else {
+        val wordBoundary = textLayoutResult.value.getWordBoundary(offset)
+        caretOffsetPosition = wordBoundary.end
+    }
+
+    onValueChange(editProcessor.toTextFieldValue().copy(selection = TextRange(caretOffsetPosition)))
+}
+
+private fun TextLayoutResultProxy.isPunctuationOrSpaceBeforeWordTapped(
+    caretOffset: Int,
+    currentText: String
+): Boolean {
+    /* From TextLayoutResultProxy.value.getWordBoundary(caretOffset) it is clear
+    * that for whitespace or punctuation mark method will return empty range.
+    * Thus, if empty range was returned, and the caretOffset greater than start and less than last index of text
+    * then offset should be somewhere between words.
+    * */
+    val char = currentText[caretOffset]
+    val isGreaterThanFirst = caretOffset >= 0
+    val isLessThanLast = caretOffset <= currentText.lastIndex
+    return char.isPunctuationOrWhitespace() && isGreaterThanFirst && isLessThanLast
+}
+
+private fun TextLayoutResultProxy.isFirstLetterOfWordTapped(caretOffset: Int): Boolean {
+    return value.getWordBoundary(caretOffset).start == caretOffset
+}
+
+private fun TextLayoutResultProxy.isCaretTapped(
+    caretOffset: Int,
+    previousCaretOffset: Int
+): Boolean {
+    return previousCaretOffset == caretOffset
+}
+
+private fun TextLayoutResultProxy.isLeftEdgeTapped(caretOffset: Int): Boolean {
+    val lineNumber = value.getLineForOffset(caretOffset)
+    val lineStartOffset = value.getLineStart(lineNumber)
+    return lineStartOffset == caretOffset
+}
+
+private fun TextLayoutResultProxy.isRightEdgeTapped(caretOffset: Int): Boolean {
+    val lineNumber = value.getLineForOffset(caretOffset)
+    val lineStartOffset = value.getLineEnd(lineNumber)
+    return lineStartOffset == caretOffset
+}
+
+private fun findNextWordBoundary(
+    caretOffset: Int,
+    currentText: String,
+    textLayoutResult: TextLayoutResultProxy
+): TextRange {
+    val wordRange = textLayoutResult.value.getWordBoundary(caretOffset)
+    val currentChar = currentText[caretOffset]
+    return if (!currentChar.isPunctuationOrWhitespace()) {
+        wordRange
+    } else if (caretOffset >= currentText.lastIndex) {
+        TextRange.Zero
```

Comment:
```suggestion
        TextRange(currentText.lastIndex + 1)
```

Otherwise If we have last char as punctuation, we'll jump to the start of the textfield, tapping on this char

Reply:
done
made just `lastIndex` because `lastIndex + 1` made a bug when textfield contains new line after last character. Without `+1` works just fine